hass_nabucasa/files.py

```diff
@@ -0,0 +1,210 @@
+import logging
+from typing import TYPE_CHECKING, Any, Literal, TypedDict
+
+from aiohttp import (
+    ClientError,
+    ClientResponse,
+    ClientTimeout,
+    StreamReader,
+    hdrs,
+)
+
+from .auth import CloudError
+
+if TYPE_CHECKING:
+    from . import Cloud, _ClientT
+
+_LOGGER = logging.getLogger(__name__)
+
+_FILE_TRANSFER_TIMEOUT = 43200.0  # 43200s == 12h
+
+type StorageType = Literal["backup"]
+
+
+class FilesError(CloudError):
+    """Exception raised when handling files."""
+
+
+class _FilesHandlerUrlResponse(TypedDict):
+    """URL Response from files handler."""
+
+    url: str
+
+
+class FilesHandlerDownloadDetails(_FilesHandlerUrlResponse):
+    """Download details from files handler."""
+
+
+class FilesHandlerUploadDetails(_FilesHandlerUrlResponse):
+    """Upload details from files handler."""
+
+    headers: dict[str, str]
+
+
+class Files:
+    """Class to help manage files."""
+
+    def __init__(
+        self,
+        cloud: Cloud[_ClientT],
+    ) -> None:
+        """Initialize cloudhooks."""
+        self._cloud = cloud
+
+    def _do_log_response(
+        self,
+        resp: ClientResponse,
+        data: list[Any] | dict[Any, Any] | str | None = None,
+    ) -> None:
+        """Log the response."""
+        isok = resp.status < 400
+        target = (
+            resp.url.path if resp.url.host == self._cloud.servicehandlers_server else ""
+        )
+        _LOGGER.log(
+            logging.DEBUG if isok else logging.WARNING,
+            "Response from %s%s (%s) %s",
+            resp.url.host,
+            target,
+            resp.status,
+            data["message"]
+            if not isok and isinstance(data, dict) and "message" in data
+            else "",
+        )
+
+    async def __call_files_api(
+        self,
+        *,
+        method: str,
+        path: str,
+        jsondata: dict[str, Any] | None = None,
+    ) -> Any:
+        """Call cloud files API."""
+        data: dict[str, Any] | list[Any] | str | None = None
+        await self._cloud.auth.async_check_token()
+        if TYPE_CHECKING:
+            assert self._cloud.id_token is not None
+            assert self._cloud.servicehandlers_server is not None
+
+        resp = await self._cloud.websession.request(
+            method=method,
+            url=f"https://{self._cloud.servicehandlers_server}/files{path}",
+            headers={
+                hdrs.ACCEPT: "application/json",
+                hdrs.AUTHORIZATION: self._cloud.id_token,
+                hdrs.CONTENT_TYPE: "application/json",
+                hdrs.USER_AGENT: self._cloud.client.client_name,
+            },
+            json=jsondata,
+        )
+
+        with contextlib.suppress(JSONDecodeError):
+            data = await resp.json()
+
+        self._do_log_response(resp, data)
+
+        if data is None:
+            raise FilesError("Failed to parse response from files handler") from None
+
+        resp.raise_for_status()
```

Comment:
Shouldn't we raise before we try parsing the response?

Reply:
If we do that, we lose out on additional context the server provides.